File: lib/src/main/java/com/tuti/api/data/Card.java
package com.tuti.api.data;

import com.google.common.base.Objects;
import com.google.gson.annotations.SerializedName;

public class Card {
    private String name;
    @SerializedName("exp_date")
    private String expiryDate;
    @SerializedName("pan")
    private String PAN;

    public String getCardIndex() {
        return cardIndex;
    }

    public void setCardIndex(String cardIndex) {
        if (!PAN.isEmpty()) {
            this.cardIndex = PAN;
            return;
        }
        this.cardIndex = cardIndex;
    }

    @SerializedName("card_index")
    private String cardIndex;
    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getExpiryDate() {
        return expiryDate;
    }

    public void setExpiryDate(String expiryDate) {
        this.expiryDate = expiryDate;
    }

    public String getPAN() {
        return PAN;
    }

    public void setPAN(String PAN) {
        this.PAN = PAN;
        this.cardIndex = PAN;
    }

    @Override
    public String toString() {
        return "Card{" +
                "name='" + name + '\'' +
                ", expiryDate='" + expiryDate + '\'' +
                ", PAN='" + PAN + '\'' +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Card card = (Card) o;
        return Objects.equal(name, card.name) && Objects.equal(expiryDate, card.expiryDate) && Objects.equal(PAN, card.PAN);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(name, expiryDate, PAN);
    }
}

File: lib/src/main/java/com/tuti/api/data/Card.kt
package com.tuti.api.data

import kotlinx.serialization.SerialName

@kotlinx.serialization.Serializable
data class Card(
        val name: String = "",
        @SerialName("exp_date")
        val expiryDate: String = "",
        @SerialName("pan")
        val PAN: String = "",
        val mobile: String = "",
        @SerialName("card_index")
        val cardIndex: String = "",
        @SerialName("user_pubkey")
        val pubkey: String = "",
)


@kotlinx.serialization.Serializable
data class SignUpCard(
        val name: String,
        @SerialName("exp_date")
        val expDate: String,
        @SerialName("pan")
        val pan: String,
        val mobile: String,
        val password: String,
        @SerialName("user_pubkey") val pubkey: String,
) {
        init {
                require(pan.length >= 16)
                require(expDate.length == 4)
                require(mobile.length == 10)
                require(password.length >= 8 )
        }
}

@kotlinx.serialization.Serializable
data class Cards (
        val cards: List<Card> = emptyList()
)

@kotlinx.serialization.Serializable
data class UserCards (
        @SerialName("pan")
        val pan: String = "",
        @SerialName("exp_date")
        val expDate: String = "",
        @SerialName("Cards")
        val cards: List<Card> = emptyList()

){
        /**
         * a helper method to returns a user main card
         * @throws Exception when a user doesn't have any cards
         * @return [Card]
         */
        fun getMainCard(): Card {
                return if (this.pan.isNotEmpty()) {
                        Card(PAN = this.pan, expiryDate = this.expDate)
                }else if (cards.isNotEmpty()) {
                        Card(PAN = this.cards.first().PAN, expiryDate = this.cards.first().expiryDate)
                } else {
                        throw Exception("user has no cards")
                }

        }
}